doc/man1/roarfctest.1
.\" roarfctest.1:

.TH "roarfctest" "1" "July 2010" "RoarAudio" "System User's Manual: roarfctest"

.SH NAME

roarfctest \- Test frequency behavior of filters in discrete time domain

.SH SYNOPSIS

roarfctest [OPTIONS]...

.SH DESCRIPTION

This program tests libroardsp filters in discrete time domain.

.SH "OPTIONS"

.TP
\fB--rate      RATE\fR
Set sample rate

.TP
\fB--sfreq     SFREQ\fR
Start frequency

.TP
\fB--efreq     EFREQ\fR
End frequency

.TP
\fB--engage    ENGAGE\fR
This option is not documented

.TP
\fB--tmax      TMAX\fR
This option is not documented

.TP
\fB--gnuplot\fR
Output data using gnuplot(1)

.TP
\fB--rmsout    RMSOUT\fR
Set output file for RMS data

.TP
\fB--help   \-h\fR
Show this help


.SH "SEE ALSO"
\fBroartips\fR(7),
\fBlibroar\fR(7),
\fBRoarAudio\fR(7).

.SH "HISTORY"

For history information see \fBRoarAudio\fR(7).

.\" ll


ChangeLog
v. 0.3beta7 - ?
	* Added support for audio profiles
	* Updated libroarrsound
	* Added RPlay protocol to roard
	* Updates of copyright and license headers
	* Added ckport file
	* Updated some tools to get away from old, legacy
	  and security relevant function calls

v. 0.3beta6 - Thu Jun 24 2010 20:47 CEST
	Prereleases: 0: Fri Jun 11 2010 10:47 CEST;
	             1: Sat Jun 12 2010 11:07 CEST;
	             2: Wed Jun 16 2010 12:52 CEST
	* Added something like select() for VIOs
	* Fixed a lot warnings
	* disabled cmd codecfilter by default
	* Fixed error with roarinterconnect on non-OSS systems
	* Fixed some problems with minimal builds
	* Some Copyright fixes
	* much better VIO support for connections
	* much better OSS emulation
	* Updated win32 port
	* Updated RSound support
	* fixed pos calc in case of PCM->PCM converting (pr1)
	* fixed libroarpulse (pr2)

v. 0.3beta5 - Sun May 02 2010 12:41 CEST
	* added tool roardtmf
	* Support to store server address in X11's root window propertys
	* some small fixes
	* wrote a lot more libroarpulse code, some async API clients work now
	* added compbins for esddsp, artsdsp, audiooss, padsp
	* splited libroarpulse and libroarpulse-simple
	* upgraded some code to use VIO
	* more updates for Debian
	* fixed init script for Debian and Debian like systems
	* fixed some mixer stuff, a lot things work better now
	* added support for notify beeps
	* updated drivers
	* updated ROAR_{ERR,WARN,INFO,DBG}()
	* added syslog support
	* updated test suit
	* done a lot work on libroaross
	* fixed segfault in ALSA plugin
	* Added support for OSS to roarinterconnect(1)
	* Added server side RSound Emulation
	* Added libroarrsound, client side RSound Emulation
	* Added ALSA driver
	* Added RSound driver
	* Added PortAudio driver

v. 0.3beta4 - Sat Feb 06 2010 26:02 CET
	* Applyed patches for Debian
	* updated roarify to load libs of correct version
	* updated list of supported versions (complibs)
	* added header with common units
	* added support for PulseAudio Simple protocol
	* added support to libroaross to emulate DMX4Linux (write only)
	* added DSTR driver to roard
	* updated librarys' soname
	* do not link OpenSSL by default (licence incompatible?)
	* support for dynamic lib loading, plugin interface
	* support to for registering DSTR types
	* moved WinSock things out of main vio.[ch]
	* support auto-tool --prefix*=*

v. 0.3beta3 - Fri Jan 22 2010 22:17 CET
	* added some aRts/NAS/YIFF/PulseAudio tools to compbins
	* added support for comp libs to roar-config
	* added support for uni-directional streams in roarinterconnect
	* write much better support for config, including codec config
	* reduced the TCP overhead produced by speex
	* added support for speex preprocessor
	* get names of complibs from table, do not guess them
	* a lot small fixes
	* moved most of the mixer stuff out of roard into libroardsp
	* added flags immutable and enhance
	* test for signal.h
	* Corrected (DSTR) HTTP Protocol support so servers without headerflush
	  should now work, too.
	* added protocol support for icy
	* Added version numbers to lib names
	* updated manpages
	* wrote OSS emulation
	* fixed test scripts

v. 0.3beta2 - Mon Oct 05 2009 19:48 CEST
	* Added prethru flag
	* Added support to read info about client's network node
	* Ported libroaresd, libroarartsc
	* Fixed some problems with EsounD emulation
	* added support for compatibility binarys
	* added poll() support to libroarsndio
	* reject streams of unsupported codec

v. 0.3beta1 - Sat Sep 26 2009 25:50 CEST
	* incremented nummber of streams per client fron 4 to 8
	* RIFF WAVE codecfilter uses virtual streams
	* another set of changed names of parameters in roarctl
	* Interface to read byteorder of a client from roard
	* implemented a 'state of stream' var
	* cleanup, small fixes
	* added VIO interface to transcoder interface
	* ask codecfilter of parent in case of new child stream
	* added 'subsystem' complex: multi subsystem data on single streams
	* added basic /Radio Data and Trasmitter Control System/.
	* better support for stereo upmixing
	* make audio group a optional thing
	* better support for win32
	* wrote WMM driver
	* better multi protocol support
	* Support for server side EsounD emulation on win32

v. 0.3beta0 - Sun Sep 06 2009 22:42 CEST
	* added new stream directions for MIDI and light control
	* cleanup of client stream struct
	* added basic light control subsystem
	* added DMX512 driver
	* wrote basic MIDI subsystem
	* added flag 'mute'
	* added flag 'mmap'
	* added lib libroarmidi and libroarlight
	* added OpenSLP support
	* rewote fillmixbuer2 code
	* added a roar_conv2() with a more clear API
	* added support to write pidfile
	* added support for encoding 32 bit Vorbis streams
	* added transcoding features in libroardsp
	* added tool 'roarphone'
	* added tool 'roarinterconnect'
	* added simple raw data subsystem
	* added working support for THRU streams
	* changed names of parameters in roarctl
	* added lib libroareio - Extended IO for RoarAudio
	* Update order of default drivers, now it is: oss, ao, esd, null
	* cleanup of old sysio driver interface
	* added server side EsounD emulation
	* added multi homed support to server
	* fixed a long outstanding bug with pa_path_get_filename()

v. 0.2 - Thu May 21 2009 21:00
	* do a some more testings for functions not existing on win32
	* init winsock
	* done a lot fixes in order to get libroar and some clients work on win32
	* done a lot cleanup on configure script

v. 0.2rc2 - Sun May 17 2009 25:44
	* corrected check for GNU peercred
	* added basic logging support to configure
	* fixed plugin configures
	* added and fixed semi-install target on plugins
	* call plugin configures from main configure
	* include plugins in Makefile known to work from configure
	* fixed Makefiles of plugins: added -I../../include, -L../../lib
	* Make use of pkg-config in plugin configures
	* corrected possition of -ls on linker commands in most Makefiles

roarclients/roarfctest.c
//roarfctest.c:

/*
 *      Copyright (C) Philipp 'ph3-der-loewe' Schafft - 2008-2010
 *
 *  This file is part of roarclients a part of RoarAudio,
 *  a cross-platform sound system for both, home and professional use.
 *  See README for details.
 *
 *  This file is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 3
 *  as published by the Free Software Foundation.
 *
 *  RoarAudio is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this software; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 51 Franklin Street, Fifth Floor,
 *  Boston, MA 02110-1301, USA.
 *
 */

#include <roaraudio.h>  /* libroar */

#ifdef ROAR_HAVE_LIBM

#include <math.h>       /* sin() */
#include <stdio.h>      /* *printf*() */
#include <libroardsp/libroardsp.h>

void usage (void) {
 printf("roarfctest [OPTIONS]...\n");

 printf("\nOptions:\n\n");

 printf(
        "  --rate      RATE      - Set sample rate\n"
        "  --sfreq     SFREQ     - Start frequency \n"
        "  --efreq     EFREQ     - End frequency\n"
        "  --engage    ENGAGE    - This option is not documented\n"
        "  --tmax      TMAX      - This option is not documented\n"
        "  --gnuplot             - Output data using gnuplot(1)\n"
        "  --rmsout    RMSOUT    - Set output file for RMS data\n"
        "  --help   -h           - Show this help\n"
       );
}

int main (int argc, char * argv[]) {
 int rate     = ROAR_RATE_DEFAULT;
 int bits     = 16;
 int channels = 1; /* mono */
 int codec    = ROAR_CODEC_DEFAULT;
 float freq   = 523.2;            /* middle C */
 float step; //   = M_PI*2*freq/rate; /* how much time per sample we have to encode ... */
 float t      = 0; /* current time */
 float length;
 int16_t amp  = 32767;
 int      i;
 char   * k;
 int32_t  tmp;
 int16_t  out[1024];
 int16_t  maxval, maxval_in;
 uint32_t sc, sc_tot = 0;
 int      count = 0;
 float    rms, rms_in = amp/sqrt(2);
 float    worklen;
 float    sfreq  = 1, efreq = -1;
 float    engage = 0;
 float    tmax   = 1024;
 struct roardsp_filterchain fc[1];
 struct roardsp_filter      filt[3];
 struct roar_stream         stream[1];
 FILE * rmsout = NULL;
 int kill_var;
#ifdef ROAR_HAVE_BIN_GNUPLOT
 int do_gnuplot = 0;
#endif

 for (i = 1; i < argc; i++) {
  k = argv[i];
  kill_var = 1;

  if ( !strcmp(k, "--rate") ) {
   rate = atoi(argv[++i]);
  } else if ( !strcmp(k, "--sfreq") ) {
   sfreq = atof(argv[++i]);
  } else if ( !strcmp(k, "--efreq") ) {
   efreq = atof(argv[++i]);
  } else if ( !strcmp(k, "--engage") ) {
   engage = atof(argv[++i]);
  } else if ( !strcmp(k, "--tmax") ) {
   tmax = atof(argv[++i]);
#ifdef ROAR_HAVE_BIN_GNUPLOT
  } else if ( !strcmp(k, "--gnuplot") ) {
   do_gnuplot = 1;
#endif
  } else if ( !strcmp(k, "--rmsout") ) {
   rmsout = fopen(argv[++i], "w");
  } else if ( !strcmp(k, "--help") || !strcmp(k, "-h") ) {
   usage();
   return 0;
  } else {
   kill_var = 0;
  }

  if ( kill_var )
   *k = 0;
 }

#ifdef ROAR_HAVE_BIN_GNUPLOT
 if ( do_gnuplot && rmsout == NULL ) {
  if ( (rmsout = popen(ROAR_HAVE_BIN_GNUPLOT, "w")) != NULL ) {
   fprintf(rmsout, "set grid\n");
   fprintf(rmsout, "set log x 10\n");
   fprintf(rmsout, "set log y 10\n");
   fprintf(rmsout, "plot \"-\" with lines title \"filter amplification\"\n");
  }
 }
#endif

 if ( roar_stream_new(stream, rate, channels, bits, codec) == -1 ) {
  pclose(rmsout);
  return 2;
 }

 if ( roardsp_filter_init(filt, stream, ROARDSP_FILTER_DCBLOCK) == -1 ) {
  ROAR_ERR("main(*): roardsp_filter_init() failed: errno=%s(%i)", strerror(errno), errno);
  pclose(rmsout);
  return 1;
 }

/*
 freq = 1000;
 roardsp_filter_ctl(filt, ROARDSP_FCTL_FREQ, &freq);
*/

 if ( roardsp_filter_init(filt+1, stream, ROARDSP_FILTER_HIGHP) == -1 ) {
  ROAR_ERR("main(*): roardsp_filter_init() failed: errno=%s(%i)", strerror(errno), errno);
  pclose(rmsout);
  return 1;
 }

 freq = 1000;
 roardsp_filter_ctl(filt+1, ROARDSP_FCTL_FREQ, &freq);

 if ( roardsp_filter_init(filt+2, stream, ROARDSP_FILTER_AMP) == -1 ) {
  ROAR_ERR("main(*): roardsp_filter_init() failed: errno=%s(%i)", strerror(errno), errno);
  pclose(rmsout);
  return 1;
 }

 tmp = 2;
 roardsp_filter_ctl(filt+2, ROARDSP_FCTL_DIV, &tmp);
 tmp = 3;
 roardsp_filter_ctl(filt+2, ROARDSP_FCTL_MUL, &tmp);


 roardsp_fchain_init(fc);
 roardsp_fchain_add(fc, filt);
// roardsp_fchain_add(fc, filt+1);
// roardsp_fchain_add(fc, filt+2);

 fprintf(stderr, "Starting analysis in frequency domain...\n");

 if ( efreq == -1 )
  efreq = (float)rate/2;

 length = 5/sfreq;

 for (freq = sfreq; freq < efreq; freq *= (1+exp(1)/100), length /= (1+exp(1)/100)) {
  step      = M_PI*2*freq/rate;
  maxval    = -amp;
  maxval_in = -amp;
  sc        = 0;
  rms       = 0;
  t         = -engage;

  if ( roardsp_fchain_reset(fc, ROARDSP_RESET_STATE) == -1 ) {
   ROAR_ERR("Can not reset filterchain.");
   return 8;
  }


  worklen = length;

  if ( worklen > tmax )
   worklen = tmax;

  worklen = 2*M_PI*freq*worklen;


  while (t < worklen) {
   for (i = 0; i < 1024; i++) {
    out[i] = amp*sin(t);
    if ( out[i] > maxval_in )
     maxval_in = out[i];
    t += step;
   }

   roardsp_fchain_calc(fc, out, 1024);

   if ( t >= 0 ) {
    for (i = 0; i < 1024; i++) {
     rms += out[i] * out[i];
     if ( out[i] > maxval )
      maxval = out[i];
    }
    sc += 1024;
   } else {
    sc_tot += 1024;
   }
  }

  count++;
  sc_tot += sc;

  rms /= sc;
  rms  = sqrt(rms);
//  rms /= amp;
  rms /= rms_in;

  printf("%f: %f %u %f\n", freq, (float)maxval/maxval_in, sc, rms);

  if ( rmsout != NULL )
   fprintf(rmsout, "%f: %f\n", freq, rms);
 }


 fprintf(stderr, "Finished analysis in frequency domain: done tests on a total of %u samples on %i frequencies\n",
           sc_tot, count);

 roardsp_fchain_uninit(fc);

 if ( rmsout != NULL )
  pclose(rmsout);

 return 0;
}

#else
int main (void) {
 fprintf(stderr, "Error: No Math library support compiled in.\n");
 return 1;
}
#endif

//ll
